core/state_transition.go

```diff
@@ -766,12 +860,24 @@ func (st *stateTransition) refundGas(refundQuotient uint64) uint64 {
 
 	st.gasRemaining += refund
 
-	// Gas refund now handled in `distributeTxFees`
-
 	// Also return remaining gas to the block gas counter so it is
 	// available for the next transaction.
 	st.gp.AddGas(st.gasRemaining)
 
+	// If the transaction uses a fee currency then the refund will be done in the call to CreditFees.
+	if st.msg.FeeCurrency != nil {
+		return refund
+	}
+
+	// Return ETH for remaining gas, exchanged at the original rate.
+	remaining := uint256.NewInt(st.gasRemaining)
+	remaining.Mul(remaining, uint256.MustFromBig(st.msg.GasPrice))
+	st.state.AddBalance(st.msg.From, remaining, tracing.BalanceIncreaseGasReturn)
+
+	if st.evm.Config.Tracer != nil && st.evm.Config.Tracer.OnGasChange != nil && st.gasRemaining > 0 {
+		st.evm.Config.Tracer.OnGasChange(st.gasRemaining, 0, tracing.GasChangeTxLeftOverReturned)
```

Comment:
This tracer call will be missing for fee-currency transactions, since the `CreditFees` does not call it